src/mlpack/core/data/normalize_labels_impl.hpp

```diff
@@ -39,32 +40,32 @@ void NormalizeLabels(const RowType& labelsIn,
   // we'll resize it back down to its actual size.
   mapping.set_size(labelsIn.n_elem);
   labels.set_size(labelsIn.n_elem);
+  // Map for mapping labelIn to their label
+  std::unordered_map<eT, size_t> labelMap;
   size_t curLabel = 0;
   for (size_t i = 0; i < labelsIn.n_elem; ++i)
   {
-    bool found = false;
-    for (size_t j = 0; j < curLabel; ++j)
+    // If labelsIn[i] is already in the map, use the existing label.
+    if (labelMap.count(labelsIn[i]) > 0)
     {
-      // Is the label already in the list of labels we have seen?
-      if (labelsIn[i] == mapping[j])
-      {
-        labels[i] = j;
-        found = true;
-        break;
-      }
+      labels[i] = labelMap[labelsIn[i]] - 1;
     }
-
-    // Do we need to add this new label?
-    if (!found)
+    else
     {
-      mapping[curLabel] = labelsIn[i];
+      // If labelsIn[i] not there then add it to Map
+      labelMap[labelsIn[i]] = curLabel + 1;
       labels[i] = curLabel;
       ++curLabel;
     }
   }
-
   // Resize mapping back down to necessary size.
   mapping.resize(curLabel);
+  // Mapping array created with encoded labels
+  for (auto it = labelMap.begin(); it != labelMap.end(); ++it)
+  {
+    mapping[(it->second) - 1] = it->first;
+  }
+  labelMap.clear();
```

Comment:
No need to clear a temporary object; as soon as the method exits it won't make a difference.

Reply:
@rcurtin , Sorry for that, I did it a lot ago when some memory test were failing. I agree it was not so useful.